File: leetcode-submissions/2359_Find_Closest_Node_to_Given_Two_Nodes.cpp
class Solution {
public:
    void DFSForCalculatingDistanceFromSourcetoDestination(unordered_map<int, vector<int>>&adj , vector<int>&dist, int currSteps, int currNode, vector<int>&visited){
        visited[currNode]= true;
        // updating distance array 
        dist[currNode]= currSteps;
        for(auto& nbr: adj[currNode]){
          if(not visited[nbr]){
              DFSForCalculatingDistanceFromSourcetoDestination(adj , dist, currSteps+1, nbr, visited);
          }

        }

    }
    int closestMeetingNode(vector<int>& edges, int node1, int node2) {
        unordered_map<int , vector<int>>adj ;
        int n = edges.size();
        for(int i=0;i<n;i++){
            // i-> edge[i]
            if(edges[i]==-1){
                continue;
            }
            adj[i].push_back(edges[i]);

        }
        vector<int>distFromNode1(n , INT_MAX);

        vector<int>distFromNode2( n, INT_MAX);
        vector<int>visited1(n , false);
        vector<int>visited2(n , false);

        DFSForCalculatingDistanceFromSourcetoDestination(adj , distFromNode1, 0 , node1, visited1);
        DFSForCalculatingDistanceFromSourcetoDestination(adj , distFromNode2, 0 , node2, visited2);
        int minDistNode = -1 ;
        int minDistTillNow= INT_MAX ;
        for(int i=0;i<n;i++){
            int maxD = max(distFromNode1[i], distFromNode2[i]);
            if(minDistTillNow>maxD){
                minDistTillNow= maxD;
                minDistNode = i;

            }
        }
        return minDistNode;
    }   
};

File: leetcode-submissions/2359_Find_Closest_Node_to_Given_Two_Nodes.ts
function closestMeetingNode(edges: number[], node1: number, node2: number): number {
    const adj:Map<number , number[]>= new Map();
    const n :number = edges.length;
    for(let i:number=0;i<n;i++){
        if(edges[i]==-1) continue;
        if(!adj.has(i)) adj.set(i ,[]);
        adj.get(i).push(edges[i]);
        
    }
    const distFromNode1:number[] = new Array(n).fill(Number.MAX_SAFE_INTEGER);
    const distFromNode2:number[] = new Array(n ).fill(Number.MAX_SAFE_INTEGER);
    const visited1:boolean[]= new Array(n).fill(false), visited2:boolean[]= new Array(n).fill(false);
    DFS(node1 , distFromNode1, adj , 0 ,visited1 );
    DFS(node2 , distFromNode2 , adj , 0  ,visited2);
    let minDistanceNode = -1, minDistance = Number.MAX_SAFE_INTEGER;
    for(let i :number =0 ;i<n;i++){
        const maxDistance = Math.max(distFromNode1[i], distFromNode2[i]);
        if(maxDistance<minDistance){
            minDistance = maxDistance ;
            minDistanceNode = i ;
        }
    }
    return minDistanceNode ;
};
function DFS(currNode :number , dist:number[], adj:Map<number , number[]>,currStep : number , visited:boolean[]):void{
    visited[currNode] = true ;
    dist[currNode]= currStep;
    if(!adj.has(currNode)) return ;
    for(const nbr of adj.get(currNode)!){
        if(!visited[nbr]){
            DFS(nbr , dist, adj , currStep+1, visited);
        }
    }
}